src/frontend/src/scheduler/distributed/stage.rs

```diff
@@ -183,14 +181,15 @@ impl StageExecution {
     /// Starts execution of this stage, returns error if already started.
     pub async fn start(&self) {
         let mut s = self.state.write().await;
-        match &*s {
-            &StageState::Pending => {
+        let cur_state = mem::replace(&mut *s, StageState::Failed);
+        match cur_state {
+            StageState::Pending { msg_sender } => {
```

Comment:
Should we add some comments here to explain why we put the sender in `state` (or just add a note above `QueryExecution::shutdown_tx`? This cyclic reference is pretty tricky, and future developers should be warned of it.